File: railway-oriented-programming/dotnet/RopFSharp/Validation.fs
module Validation

    open Input
    open Rop
    
    let validateInput input =
        if input.name = "" then Failure "Name must not be blank"
        else if input.email  = "" then Failure "Email must not be blank"
        else Success input

    let nameIsNotEmpty input =
        if input.name = "" then Failure "Name must not be blank"
        else Success input

    let emailLengthIsLessThan50Chars input =
        if input.email.Length > 50 then Failure "Email must not be longer than 50 chars"
        else Success input

    let emailIsNotEmpty input =
        if input.email  = "" then Failure "Email must not be blank"
        else Success input

    let combinedValidation = 
        // connect the two-tracks together
        nameIsNotEmpty 
        >> bind emailLengthIsLessThan50Chars 
        >> bind emailIsNotEmpty

    // let combinedValidation = 
    //     // convert from switch to two-track input
    //     let validate2' = bind validate2
    //     let validate3' = bind validate3
    //     // connect the two-tracks together
    //     validate1 >> validate2' >> validate3'

File: railway-oriented-programming/dotnet/RopCSharp/Validation.cs
namespace RopCSharp;

internal class Validation
{
    internal Result<Input, Error> CombineValidation(Input input)
      => NameIsNotEmpty(input)
        .OnSuccess(EmailIsNotEmpty)
        .OnSuccess(EmailDoesNotExceedMaxLength)
        ;

    internal Result<Input, Error> NameIsNotEmpty(Input input)
      => string.IsNullOrWhiteSpace(input.Name)
          ? Failure(Domain.Errors.Name.IsEmpty)
          : Success(input);

    internal Result<Input, Error> EmailIsNotEmpty(Input input)
        => string.IsNullOrWhiteSpace(input.Email)
            ? Failure(Domain.Errors.Email.IsEmpty)
            : Success(input);

    internal Result<Input, Error> EmailDoesNotExceedMaxLength(Input input)
        => input.Email.Length <= 50
            ? Success(input)
            : Failure(Domain.Errors.Email.TooLong);
}